src/dotVariant.Generator/SourceGenerator.cs

```diff
@@ -14,78 +14,80 @@
-                receiver
-                .VariantDecls
-                .Select(decl => (decl.Symbol, decl.Syntax, decl.Nullable, Diags: Diagnose.Variant(decl.Symbol, decl.Syntax, context.CancellationToken)))
-                .Memoize();
-
-            decls
-                .SelectMany(desc => desc.Diags)
-                .ForEach(context.ReportDiagnostic);
-
-            Descriptors =
-                decls
-                .Where(decl => !decl.Diags.Any(d => d.Severity == DiagnosticSeverity.Error))
-                .Select(decl => Descriptor.FromDeclaration(decl.Symbol, decl.Syntax, decl.Nullable))
-                .ToImmutableArray();
-
-            RenderInfos =
-                Descriptors
-                .Select(desc => RenderInfo.FromDescriptor(desc, (CSharpCompilation)context.Compilation, context.AnalyzerConfigOptions, context.CancellationToken))
-                .ToImmutableArray();
-
-            Enumerable
-                .Zip(Descriptors, RenderInfos, (desc, ri) => (desc, ri))
-                .ForEach(v => context.AddSource(SanitizeName(v.desc.Type.ToDisplayString()), Renderer.Render(v.ri)));
-
-            static string SanitizeName(string name)
-                => name
-                // If the contains type parameters replace angle brackets as those are not allowed in AddSource()
-                .Replace('<', '{')
-                .Replace('>', '}')
-                // Escaped names like @class or @event aren't supported either
-                .Replace('@', '.');
-        }
+#if DEBUG
+        public List<RenderInfo> RenderInfos { get; } = new();
+#endif
 
-        public ImmutableArray<Descriptor> Descriptors { get; private set; }
-        public ImmutableArray<RenderInfo> RenderInfos { get; private set; }
-
-        public void Initialize(GeneratorInitializationContext context)
-        {
-            context.RegisterForSyntaxNotifications(() => new SyntaxReceiver());
-        }
-
-        private sealed class SyntaxReceiver : ISyntaxContextReceiver
+        public void Initialize(IncrementalGeneratorInitializationContext generatorContext)
         {
-            public List<(INamedTypeSymbol Symbol, TypeDeclarationSyntax Syntax, NullableContext Nullable)> VariantDecls { get; } = new();
-
-            public void OnVisitSyntaxNode(GeneratorSyntaxContext context)
+            var variantDecls = generatorContext.SyntaxProvider.CreateSyntaxProvider((node, ct) =>
+            {
+                const nuint declarationBloom = (nuint)SyntaxKind.ClassDeclaration | (nuint)SyntaxKind.StructDeclaration |
+                                             (nuint)SyntaxKind.RecordDeclaration | (nuint)SyntaxKind.RecordStructDeclaration;
+                var nodeKind = node.Kind();
+                return ((nuint)nodeKind & declarationBloom) != default && nodeKind is SyntaxKind.ClassDeclaration
+                    or SyntaxKind.StructDeclaration
+                    or SyntaxKind.RecordDeclaration or SyntaxKind.RecordStructDeclaration;
```

Comment:
Any particular reason why you're using `nuint` here?

Reply:
Casting to `nuint` prevents masking of the slot during `ldarg_n`. It is nonsensical to do this, because the performance benefit is virtually zero.
To be more readable, we can removed the entire `declarationBloom`.